substreams/ethereum-uniswap-v4/src/modules/6_map_protocol_changes.rs

```diff
@@ -0,0 +1,228 @@
+                    .nth(1)
+                    .unwrap(),
+            )
+            .unwrap();
+            let tx = change.transaction.unwrap();
+            let builder = transaction_changes
+                .entry(tx.index)
+                .or_insert_with(|| TransactionChangesBuilder::new(&tx.into()));
+
+            builder.add_entity_change(&EntityChanges {
+                component_id: change.pool_address.to_hex(),
+                attributes: vec![Attribute {
+                    name: "liquidity".to_string(),
+                    value: new_value_bigint.to_signed_bytes_be(),
+                    change: ChangeType::Update.into(),
+                }],
+            });
+        });
+
+    // Insert others changes
+    events
+        .pool_events
+        .into_iter()
+        .flat_map(event_to_attributes_updates)
+        .for_each(|(tx, pool_address, attr)| {
+            let builder = transaction_changes
+                .entry(tx.index)
+                .or_insert_with(|| TransactionChangesBuilder::new(&tx));
+            builder.add_entity_change(&EntityChanges {
+                component_id: pool_address.to_hex(),
+                attributes: vec![attr],
+            });
+        });
+
+    Ok(BlockChanges {
+        block: Some((&block).into()),
+        changes: transaction_changes
+            .drain()
+            .sorted_unstable_by_key(|(index, _)| *index)
+            .filter_map(|(_, builder)| builder.build())
+            .collect::<Vec<_>>(),
+    })
+}
+
+fn event_to_attributes_updates(event: PoolEvent) -> Vec<(Transaction, PoolAddress, Attribute)> {
+    match event.r#type.as_ref().unwrap() {
+        pool_event::Type::Swap(swap) => vec![
+            (
+                event
+                    .transaction
+                    .clone()
+                    .unwrap()
+                    .into(),
+                hex::decode(event.pool_id.clone()).unwrap(),
+                Attribute {
+                    name: "sqrt_price_x96".to_string(),
+                    value: BigInt::from_str(&swap.sqrt_price_x96)
+                        .unwrap()
+                        .to_signed_bytes_be(),
+                    change: ChangeType::Update.into(),
+                },
+            ),
+            (
+                event.transaction.unwrap().into(),
+                hex::decode(event.pool_id).unwrap(),
+                Attribute {
+                    name: "tick".to_string(),
+                    value: BigInt::from(swap.tick).to_signed_bytes_be(),
+                    change: ChangeType::Update.into(),
+                },
+            ),
+        ],
+        pool_event::Type::ProtocolFeeUpdated(sfp) => {
+            // Mask to extract the lower 12 bits (0xFFF corresponds to 12 bits set to 1)
+            let lower_12_bits = sfp.protocol_fee & 0xFFF;
+
+            // Shift right by 12 bits and mask again to get the next 12 bits
+            let upper_12_bits = (sfp.protocol_fee >> 12) & 0xFFF;
```

Comment:
Are we lucky and there is an example of this in Sepolia? Else I'm afraid it will be hard to test this logic

Reply:
I'll try to make one during cd

Reply:
But yeah - no events so far

Reply:
And - super hard bc only the protocol owner can change this one